hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -1149,7 +1149,9 @@ private class ConnectToAllClusterMembersTask implements Runnable {
 
         @Override
         public void run() {
-            if (!client.getLifecycleService().isRunning()) {
+            if (!client.getLifecycleService().isRunning() || activeConnections.isEmpty()) {
+                // do not attempt to connect members here when the client disconnects from the cluster
+                // since we want this cluster reconnection to be only performed by `doConnectToCluster()`
                 return;
             }
```

Comment:
What-if the last active connection drops right at this moment? We'll attempt to re-connect the member, while at the same time we'll trigger the cluster reconnection logic. That's what happened anyways before this fix. But is it safe?